File: observer.java
// Certainly! The Observer design pattern is used when an object, known as the subject, maintains a list of its dependents, known as observers, that need to be notified of any changes to the subject's state. The following is a complex example of the Observer design pattern implemented in Java.

// Copy
import java.util.ArrayList;
import java.util.List;

// Subject interface that defines methods for adding, removing, and notifying observers
interface Subject {
    void addObserver(Observer observer);
    void removeObserver(Observer observer);
    void notifyObservers();
}

// Concrete subject class that implements the Subject interface
class ConcreteSubject implements Subject {
    private int state;
    private List<Observer> observers = new ArrayList<>();

    public int getState() {
        return state;
    }

    public void setState(int state) {
        this.state = state;
        notifyObservers(); // Notify observers whenever the state changes
    }

    @Override
    public void addObserver(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(this);
        }
    }
}

// Observer interface that defines the update method
interface Observer {
    void update(Subject subject);
}

// Concrete observer class that implements the Observer interface
class ConcreteObserver implements Observer {
    private String name;

    public ConcreteObserver(String name) {
        this.name = name;
    }

    @Override
    public void update(Subject subject) {
        if (subject instanceof ConcreteSubject) {
            int newState = ((ConcreteSubject) subject).getState();
            System.out.println("Observer " + name + " has been notified. New state: " + newState);
        }
    }
}

public class ObserverPatternExample {
    public static void main(String[] args) {
        ConcreteSubject subject = new ConcreteSubject();

        ConcreteObserver observer1 = new ConcreteObserver("Observer 1");
        ConcreteObserver observer2 = new ConcreteObserver("Observer 2");

        // Adding observers to the subject
        subject.addObserver(observer1);
        subject.addObserver(observer2);

        // Changing the state of the subject will notify the observers
        subject.setState(10);

        // Removing observer1
        subject.removeObserver(observer1);

        // Changing the state again will notify only observer2
        subject.setState(20);
    }
}
// In this example:

// ConcreteSubject is the subject that maintains a state and a list of observers.
// ConcreteObserver is a concrete observer that implements the update method to react to changes in the subject's state.
// ObserverPatternExample demonstrates how to use the Observer pattern by creating a subject, adding observers, and changing the subject's state to see how observers are notified.
// This example showcases a simple implementation of the Observer design pattern in Java.

File: observer.go
// Certainly! Here's an example of the Observer design pattern implemented in Go:

// Copy
package main

import (
	"fmt"
	"sync"
)

// Subject interface defines methods for adding, removing, and notifying observers
type Subject interface {
	AddObserver(observer Observer)
	RemoveObserver(observer Observer)
	NotifyObservers()
}

// Observer interface defines the update method
type Observer interface {
	Update(subject Subject)
}

// ConcreteSubject struct is a concrete implementation of the Subject interface
type ConcreteSubject struct {
	state     int
	observers []Observer
	mutex     sync.Mutex
}

// AddObserver adds an observer to the list
func (s *ConcreteSubject) AddObserver(observer Observer) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	s.observers = append(s.observers, observer)
}

// RemoveObserver removes an observer from the list
func (s *ConcreteSubject) RemoveObserver(observer Observer) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	for i, obs := range s.observers {
		if obs == observer {
			s.observers = append(s.observers[:i], s.observers[i+1:]...)
			break
		}
	}
}

// NotifyObservers notifies all observers of a change
func (s *ConcreteSubject) NotifyObservers() {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	for _, observer := range s.observers {
		observer.Update(s)
	}
}

// SetState sets the state of the subject and notifies observers
func (s *ConcreteSubject) SetState(state int) {
	s.state = state
	s.NotifyObservers()
}

// ConcreteObserver struct is a concrete implementation of the Observer interface
type ConcreteObserver struct {
	name string
}

// Update implements the update method to react to changes in the subject's state
func (o *ConcreteObserver) Update(subject Subject) {
	if concreteSubject, ok := subject.(*ConcreteSubject); ok {
		newState := concreteSubject.state
		fmt.Printf("Observer %s has been notified. New state: %d\n", o.name, newState)
	}
}

func main() {
	subject := &ConcreteSubject{}

	observer1 := &ConcreteObserver{name: "Observer 1"}
	observer2 := &ConcreteObserver{name: "Observer 2"}

	// Adding observers to the subject
	subject.AddObserver(observer1)
	subject.AddObserver(observer2)

	// Changing the state of the subject will notify the observers
	subject.SetState(10)

	// Removing observer1
	subject.RemoveObserver(observer1)

	// Changing the state again will notify only observer2
	subject.SetState(20)
}
main()
// In this example:

// ConcreteSubject is the concrete implementation of the Subject interface.
// ConcreteObserver is the concrete implementation of the Observer interface.
// The main function demonstrates how to use the Observer pattern in Go by creating a subject, adding observers, and changing the subject's state to see how observers are notified.
// This example uses a mutex to synchronize access to the observers' list, ensuring thread safety when adding or removing observers.